[KDF] Add KBKDF implementation for counter-mode HMAC

frozencemetery opened the pull request:
Implement SP800-108 section 5.1 with HMAC intended for use in Kerberos.
Add test vectors from RFC 8009.

Signed-off-by: Robbie Harwood <rharwood@redhat.com>

### Notes

Hi!  First of krb5 with tested using this KDF can be found at https://github.com/krb5/krb5/compare/master...frozencemetery:800-108test .  I've also included test vectors.

I'm not really happy with the name of this thing: SP800-108 doesn't really name it, but NIST gives it the name KBKDF in https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation

Kerberos aficionados will note that there are other KDFs we use which are not present in OpenSSL.  Should it be acceptable to the project, I will add them as time allows.  However, in the interest of getting some feedback, I'm proposing what I have for now.

This implementation is not derived in any way from krb5's (beyond sharing of test vectors from RFC, of course), of which I am not the author.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented on providers/fips/fipsprov.c:
```diff
@@ -350,6 +350,7 @@ static const OSSL_ALGORITHM fips_kdfs[] = {
     { OSSL_KDF_NAME_SSKDF, "fips=yes", kdf_sskdf_functions },
     { OSSL_KDF_NAME_PBKDF2, "fips=yes", kdf_pbkdf2_functions },
     { OSSL_KDF_NAME_TLS1_PRF, "fips=yes", kdf_tls1_prf_functions },
+    { OSSL_KDF_NAME_KBKDF, "fips=yes", kdf_kbkdf_functions },
```
I don't think we should be bringing this inside the FIPS module without a very strong reason. We've already agreed the scope of the FIPS module and the algorithms in it - so I'm not sure we can ask the lab to test more than that.

slontis replied:
It is possible for this to go into the fips module - it might just be vendor affirmed (meaning the lab would not test).

simo5 replied:
@mattcaswell the reason for this work is to attempt to have krb5 be fips compliant, we contacted our lab and they determined this KDF is compliant, but we can apply downstream patches if necessary.

Also note that we will be submitting a few more KDFs (the older rfc3961 derive key, and one used by pkinit), so it is possible I will pull this work into a larger PR so it can be considered all at once.

mattcaswell commented on providers/common/kdfs/build.info:
```diff
@@ -1,4 +1,4 @@
+$COMMON=tls1_prf.c hkdf.c kbkdf.c pbkdf2.c sskdf.c
 
 LIBS=../../../libcrypto
 SOURCE[../../../libcrypto]=$COMMON
```
Add to libcrypto only (so not `$COMMON`). See my other comment.

paulidale replied:
Kind of agreed.

I'm thinking that a Kerberos provider might be desirable -- a question yet to be discussed.

mattcaswell commented:
I've not properly reviewed this - just added my thoughts about this in relation to the FIPS module.

richsalz commented:
I also agree that it's probably too late to put this into the FIPS module, but *HOLY COW* this is an impressive PR in terms of getting all the FIPS/provider/PARAM stuff right!

levitte commented:
OK, one thing first: wow, an externally submitted provider implementation of, well, anything! I hadn't expected this.

Second, I'm a bit surprised by the `crypto/objects` change. It shouldn't be necessary.

slontis commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+ EVP_KDF *kdf;
+ EVP_KDF_CTX *kctx;
+ unsigned char out[10];
+ OSSL_PARAM params[4], *p = params;
```
params[5]

slontis commented on test/evp_kdf_test.c:
```diff
@@ -297,6 +297,95 @@ static int test_kdf_x963(void)
+    params[i] = OSSL_PARAM_construct_end();
+
+    kctx = get_kdfbyname(SN_kbkdf);
+    ret &= TEST_ptr(kctx);
```
please do as a if...

paulidale replied:
As it stands, all the following calls are executed when the first fails.  This isn't good.

frozencemetery replied:
@paulidale They shouldn't be - they'll short-circuit out.  I'll fix it to comply with @slontis's wishes though.

richsalz replied:
No, ```&=``` doesn't short circuit; I think you are confusing it with ```&&``` which does.

richsalz replied:
```
; cat ss.c ; gcc ss.c ; ./a.out

#include <stdio.h>

static int foo()
{
    printf("foo called\n");
    return 1;
}

int main() {
    int x = 0;
    x &= foo();
    return x;
}
foo called
; 
```

paulidale replied:
Thanks @richsalz

slontis commented on test/evp_kdf_test.c:
```diff
@@ -297,6 +297,95 @@ static int test_kdf_x963(void)
+    params[i] = OSSL_PARAM_construct_end();
+
+    kctx = get_kdfbyname(SN_kbkdf);
+    ret &= TEST_ptr(kctx);
```
same here

slontis commented on test/evp_kdf_test.c:
```diff
@@ -297,6 +297,95 @@ static int test_kdf_x963(void)
+static int test_kdf_kbkdf(void)
+{
+    int ret = 1;
+    ret &= test_kdf_kbkdf_8009_prf1();
```
Just make this 2 seperate test cases..

paulidale replied:
Agreed, two `ADD_TEST` lines below is better.

slontis commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM params[3];
+    unsigned char *K_i = NULL;
+    uint32_t L = htobe32(keylen * 8);
```
Is this going to work across all versions (There are no other references to this currently in the codebase - it seems to come from endian.h on my local machine).

paulidale replied:
Agreed, our usual endian handling _union_ is more appropriate.
Not that I particularly like it but it is invasive throughout the source code.

Personally, I'd not worry about the _inline_.   A static function is as good these days.

paulidale commented:
Great effort!

slontis commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    size_t KI_len;
+    unsigned char *Label;
+    size_t Label_len;
+    unsigned char *Context;
```
Not sure this uppercase naming is commonly used in structs.

paulidale replied:
I'd prefer lower case here like @slontis.  I'd cope with `KI` but would prefer `Ki` or `k_i` (this might clash later in the file).

frozencemetery replied:
I see a preference for lowercased names, so I'll change to those.

slontis commented:
> Great effort!

Yep its pretty clean..

slontis commented:
I assume you have a leak somewhere (looking at the travis errors)

slontis commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+ else if (EVP_KDF_derive(kctx, out, sizeof(out)) <= 0)
+     error("EVP_KDF_derive");
+
+ EVP_KDF_CTX_free(kctx);
```
EVP_KDF_free(kdf);

paulidale replied:
No, this is correct.

The `EVP_KDF_free()` is called right after the `EVP_KDF_CTX_new()` above.  The absence of error checking there is my fault I suspect.

frozencemetery replied:
Yeah, I modeled this after EVP_KDF-SS.pod which has similar setup.

slontis commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+done:
+    if (ret != 1)
+        OPENSSL_cleanse(key, keylen);
+    EVP_MAC_CTX_free(ctx_init);
```
EVP_MAC_free(mac); will be required if you do a fetch

paulidale replied:
Yes for this one.  It could be done earlier.

levitte commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+ EVP_KDF_free(kdf);
+
+ *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
+                                         SN_sha256, strlen(SN_sha256));
```
Also, generally speaking, I would avoid using the `SN_` macros and rather use an actual string, `"SHA256"`.
Sure, `SN_sha256` exists, but if someone wants to use this with an algorithm (let's call it "BLARG") that we don't know about but that's provided with some other provider module, there isn't going to be a corresponding `SN_blarg`, and we don't want to get the "where's the `SN_blarg` macro?" kind of issue in the future.

paulidale replied:
Instead of `strlen(3)`, sizeof works and adds the termination byte.

As an aside: this is one thing that has stumbled me a few times thus far....

paulidale commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+
+=item B<OSSL_KDF_PARAM_PROPERTIES> ("properties") <UTF8 string>
+
+=item B<OSSL_KDF_PARAM_DIGEST> ("digest") <UTF8 string>
```
Is "mac" appropriate here too?
I think SSKDF uses one.

frozencemetery replied:
SSKDF does use the MAC parameter, but for toggling between HMAC and KMAC.  This makes sense for KBKDF as well since SP800-108 also defines CMAC, but I didn't implement that.  Were we to do so, we would use MAC for toggling between the two.

Unless I'm missing something about the interface, of course.

frozencemetery replied:
If you're suggesting that we add the parameter now, I also wouldn't be opposed to that.

paulidale replied:
I was suggesting adding the mac parameter.

paulidale replied:
If adding the CMAC flavour is a lot of work, it could be done a separate PR and the MAC option would make sense there.

frozencemetery replied:
I've added the MAC parameter.  I'm not sure how much work CMAC would be, but I didn't do it because I don't have a use for it.  (krb5 with camellia does use CMAC, but it's with feedback mode, not counter mode - which presumably would be another toggle, along with double-pipeline if that were implemented.)

paulidale commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+ EVP_KDF_CTX *kctx = EVP_KDF_CTX_new(kdf);
+
+The output length of an KBKDF is specified via the C<keylen>
+parameter to the L<EVP_KDF-derive(3)> function.
```
@levitte should this be `EVP_KDF(3)-derive`?
I'm not sure I yet understand the differences in naming completely.

levitte replied:
Uhmmm, no, this should be `L<EVP_KDF_derive(3)>`

As for what you're asking...

`EVP_KDF_derive` is an existing function, yeah?  So to get the information, you would do `man EVP_KDF_derive`.

However, `EVP_KDF-KB` indicates that this isn't about an actual symbol (as `EVP_KDF_KB` might make you think), but something else.  The neat thing with dashes for this is that `man` will attempt to find a man-page for multiple arguments by concatenating them with dashes in between, so `man EVP_KDF KB` is equivalent to `man EVP_KDF-KB` and becomes a neat way to express that this is an `EVP_KDF` named "KB".

kaduk replied:
My FreeBSD man(1) does not have this behavior; multiple arguments are treated as independent manuals to display in sequence (after close of pager)

levitte replied:
It's quite possible the concatenation is a Linux thing.  However, nothing stops you from typing `man EVP_KDF-SSHKDF`, it works as well on all platforms, as expected.

However, the fact remains that man3 manuals should be named the same way as the functions it describes.  Since the function in question is called `EVP_KDF_derive`, the manual should be `EVP_KDF_derive(3)`.  That is, unless we can suddenly use dashes in C symbols :wink:

(incidently, perl6 supports dashes and apostrophes in symbol names...  that'll make for interesting page references :laughing:)

paulidale commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+
+=head1 HISTORY
+
+This functionality was added to OpenSSL FUTURE_RELEASE.
```
`FUTURE_RELEASE` ?????

levitte replied:
Well, @frozencemetery doesn't *know* if we're going to merge it with 3.0 or some time later.  It seems like a cautious move.

frozencemetery replied:
Yep.  (If I'm being totally honest, it's also because I can't quite figure out how the OpenSSL release numbering scheme works - I would guess that the next candidates are 1.2.0 or 1.1.1e?)

levitte replied:
Next candidate is 3.0

frozencemetery replied:
Thanks!

mattcaswell replied:
This might help:
https://www.openssl.org/blog/blog/2018/11/28/version/

paulidale commented on providers/common/kdfs/hkdf.c:
```diff
@@ -14,6 +14,7 @@
 #include <openssl/evp.h>
 #include <openssl/kdf.h>
 #include <openssl/core_names.h>
+#include <openssl/params.h>
```
Why?  This is the only change to this file.  Is there a case where it is required?
I'd not be surprised if I missed something....

paulidale commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+
+static void kbkdf_free(void *vctx)
+{
+    KBKDF *ctx = vctx;
```
Personally, I'd cast this to appease C++ compilers.
However, consider this optional.

paulidale commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+        if (ctx == NULL)
+            goto done;
+
+        if (!EVP_MAC_update(ctx, (unsigned char *)&i, 4) ||
```
The official syntax is:

```
if (cond1
    || cond2
    || con3)
```

An extra indentation is permitted before the `||`'s but is not essential.

frozencemetery replied:
Okay.  Might be worth adding a note to https://www.openssl.org/policies/codingstyle.html since it's not mentioned there.

paulidale commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+            !EVP_MAC_update(ctx, (unsigned char *)&L, 4))
+            goto done;
+
+        if (!EVP_MAC_final(ctx, K_i, NULL, h))
```
This should be merged into the previous _if_.

paulidale commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+
+    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, provctx))
```
Should the MAC used be set here too?
I'm happy with always using HMAC like SSKDF, so consider this a naïve question.

levitte replied:
Change the `ossl_prov_digest_load_from_params` line to:

```
    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DIGEST);
    if (p != NULL && !kbkdf_set_string(&ctx->digest, p))
        return 0;

    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_PROPERTIES);
    if (p != NULL && !kbkdf_set_string(&ctx->properties, p))
        return 0;
```

paulidale commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    } 
+    mdname = EVP_MD_name(md);
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
```
It's probably my fault, but this doesn't seem like a great place to be building and passing OSSL_PARAM arrays.

frozencemetery replied:
Can you elaborate on what you mean here?

paulidale replied:
The lookup and setup can be done in the set_params function.  It doesn't need to be delayed to the derive.

paulidale commented on test/evp_kdf_test.c:
```diff
@@ -297,6 +297,95 @@ static int test_kdf_x963(void)
+    EVP_KDF_CTX_free(kctx);
+    return ret;
+}
+static int test_kdf_kbkdf_8009_prf2(void)
```
NIT: blank line above.

paulidale commented on test/evp_kdf_test.c:
```diff
@@ -297,6 +297,95 @@ static int test_kdf_x963(void)
+    EVP_KDF_CTX *kctx;
+    OSSL_PARAM params[5];
+    char *label = "prf", *digest = "sha384", *prf_input = "test";
+    unsigned char input_key[] = {
```
I'd make these `static const`, but it isn't very important.

frozencemetery replied:
That would require a cast later on, so I'm disinclined to do that.

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+{
+    const OSSL_PARAM *p;
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
```
Rename `provctx` to `libctx`, please

levitte commented on doc/man7/EVP_KDF-KB.pod:
```diff
@@ -0,0 +1,111 @@
+L<EVP_KDF-CTX_free(3)>,
+L<EVP_KDF-ctrl(3)>,
+L<EVP_KDF-size(3)>,
+L<EVP_KDF-derive(3)>,
```
Change all dashes to underscores

paulidale replied:
And it is `L<EVP_KDF(3)>` at the top of the list :)

richsalz replied:
Should find-doc-nits complain if it finds a dash in an ```L<>``` reference?
I notice that there are files with matching "*-*.pod" :(

levitte replied:
No. Some of them are correct.
Or... If you restrict yourself to section 3 manuals, then yes.

paulidale commented:
My remaining question is: _in what way was the documentation and existing code deficient?_
We want the community to be able to write provider based algorithms...

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+        return 0;
+    }
+
+    sz = EVP_MD_size(md);
```
This, I don't quite understand...  when reading the spec, what I see is this:

> Return: *K<sub>O</sub>, i.e., the leftmost *L* bits of *`result(n)`*. 

Ref: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+/* Our context structure. */
+typedef struct {
+    void *provctx;
+    PROV_DIGEST digest;
```
Change `PROV_DIGEST digest;` to:

``` C
    char *digest;
    char *properties;
```

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+{
+    KBKDF *ctx = vctx;
+
+    ossl_prov_digest_reset(&ctx->digest);
```
Change this to:

``` C
    OPENSSL_free(ctx->digest);
    OPENSSL_free(ctx->properties);
```

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    int ret = 0;
+    const EVP_MD *md;
+    EVP_MAC *mac;
+    const char *mdname;
```
Remove `const char *mdname`

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    } 
+    mdname = EVP_MD_name(md);
```
Remove all lines from 163 to here.  You can move the check of `KI` and `KI_len` down after setting the params of the MAC, 'cause at that point, you know you have a MAC to play with.

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,303 @@
+    mdname = EVP_MD_name(md);
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+                                                 (char *)mdname, 0);
```
Use `ctx->digest` instead of `(char *)mdname`

levitte replied:
Also, make sure to add a properties param...

levitte commented:
@frozencemetery, I hope you're coping with all the feedback!  If nothing else, you sure have attracted out attention!

frozencemetery commented:
Hi, thanks for your feedback!  (And your patience while I work on addressing it; I'll have a new revision soon.)

@paulidale asked:

> My remaining question is: in what way was the documentation and existing code deficient?
> We want the community to be able to write provider based algorithms...

Well, to be honest, I didn't read documentation on this because I couldn't find any.  I was mostly modeling this on what sskdf and hkdf do (and occasionally the other KDFs).  There are a couple of... odd... mistakes that came out of this, which hopefully have all now been caught by this round of review.

The two hardest parts for me were the magic symbols (i.e., finding what files to edit to get `make update` to make the right `#define`s and such) and coping with changes to the interface while I was writing this (I think you can see a lot of things you've fixed elsewhere recently that I hadn't here because they didn't visibly break when done "the old way"). :)

simo5 commented:
@levitte one thing that I do not understand about the providers is why you have the common/ and default/ directories in there.
The problem is if you have stuff in default/ you cannot add it to the fips stuff?
Why is there a difference?
(You'll understand why I ask when I submit my next PR in a few minutes)

mattcaswell commented:
> one thing that I do not understand about the providers is why you have the common/ and default/ directories in there.

The idea was that there would be a directory for each provider, so one for fips, one for legacy and one for default. Code that is specific to each provider would go into the appropriate directory. Some code is shared between multiple providers (most notably the fips algorithm implementations are largely a subset of the default algorithm implementations) and so they go in common.

Personally, I think that where we have code that is a candidate for sharing across multiple providers I don't have a big problem with it going into common - even if it is currently only used by one.

@slontis says above that we might be able to include this in the FIPS module anyway without the lab testing it (vendor affirmed). I'm not FIPS savvy enough to know what the implications of that are. I'd be interested in @t-j-h's opinion.

levitte commented:
@frozencemetery, regarding my comments on how to treat underlying MAC algos, you might want to look at what has happened with SSKDF...  there' been some development that in this regard, which also supercede what I've said on the subject here.

Sorry for being a bother...

levitte commented:
There's one comment made by @simo5 that kind of caught me off guard:

> @mattcaswell the reason for this work is to attempt to have krb5 be fips compliant, we contacted our lab and they determined this KDF is compliant, but we can apply downstream patches if necessary.
> 
> Also note that we will be submitting a few more KDFs (the older rfc3961 derive key, and one used by pkinit), so it is possible I will pull this work into a larger PR so it can be considered all at once.

This makes it all sound like this effort is really a bit of https://github.com/krb5/krb5 that somehow becomes part of OpenSSL, because...  why?  For whose benefit?  And especially revolving around FIPS and validation, how do you plan on getting this piece of code validated if you have a validation mechanism of your own?  (not that I fully understand the ins and outs of validation, but this does strike me as a bit tangled)

We have designed the whole provider machine to make it possible for other parties to have their own providers in the own source and installation and have that pass through libcrypto, or even to have a provider that's part of an application and have it treated as a built-in provider (see the manual for `OSSL_PROVIDER_add_builtin`, i.e. `doc/man3/OSSL_PROVIDER.pod`).

It's not that we dislike contributions like this, far from it (you did notice the initial welcome it got, I hope?), but I'd like also to encourage a rational look at what that means long term in terms of maintenance.  If you look at it from that angle, is it better that in lives in OpenSSL's source, or in krb5's?

simo5 commented:
> This makes it all sound like this effort is really a bit of https://github.com/krb5/krb5 that somehow becomes part of OpenSSL, because... why? For whose benefit?

We've been trying for a while (at Red Hat) to move all the cryptography used in our krb5 packages to use openssl's crypto. One of the reasons is to be able to affirm krb5 to be FIPS compliant by using the validated openssl crypto module.
In order to do that all algorithms FIPS consider their purview need to be in the validated crypto module, otherwise you end up having to validate the krb5 "module" as well.
Who benefits is anyone that wants to run krb5 in a FIPS compliant environemnt using openssl as the validated crypto module. 

> And especially revolving around FIPS and validation, how do you plan on getting this piece of code validated if you have a validation mechanism of your own? (not that I fully understand the ins and outs of validation, but this does strike me as a bit tangled)

I am not sure I fully understand the premise of your question given any vendor that distributes binaries of crypto modules needs to perform validation, as FIPS validation certifies binaries not source code. In our case we, of course, intend to validate Openssl as our crypto module than vendor affirm FIPS compliance of the krb5 implementation because all FIPS relevant algorithms are validated in the crypto module.

> We have designed the whole provider machine to make it possible for other parties to have their own providers in the own source and installation and have that pass through libcrypto, or even to have a provider that's part of an application and have it treated as a built-in provider (see the manual for `OSSL_PROVIDER_add_builtin`, i.e. `doc/man3/OSSL_PROVIDER.pod`).

Sure, we can definitely patch the fips provider downstream and add there what are the FIPS algorithm we end up certifying. If you think that's a better course of action we can simply not mark these KDFs as fips, we still prefer to upstream all the KDFs so that, even if not formally validated, the upstream krb5 project can also be contributed code that uses KDFs from OpenSSL without extensive downstream patching of multiple projects.

> It's not that we dislike contributions like this, far from it (you did notice the initial welcome it got, I hope?), but I'd like also to encourage a rational look at what that means long term in terms of maintenance. If you look at it from that angle, is it better that in lives in OpenSSL's source, or in krb5's?

It really is a matter of perspective, however having these KDFs in OpenSSL allows everyone to use them regardless of whether they use MIT's krp5, Heimdal's Krb5, or some other custom implementation of Krb5, and keep FIPS validation in a single component.

mattcaswell commented:
This all makes a great deal of sense to me. 

> In our case we, of course, intend to validate Openssl as our crypto module than vendor affirm FIPS compliance of the krb5 implementation because all FIPS relevant algorithms are validated in the crypto module.

So, IIUC, you will do your own validation based on our module code as a starting point.

From my point of view OpenSSL is supposed to be a general purpose crypto library. So if there are standards based KDFs out there that are being used in the real world, then there really is no reason why they shouldn't have a home in OpenSSL (and lots of reasons why they should).

It was always our intention that our code can be the starting point for other people's validations. The only question really is whether these KDF's should be in our *own* module or not. If the answer is "yes" then I guess that means us taking on the costs/paperwork etc of getting it validated as part of our module. If the answer is "no" then we should make it as easy as possible for downstream users to pull that code into the FIPS module if they wish to do so.

mattcaswell commented:
> If the answer is "no" then we should make it as easy as possible for downstream users to pull that code into the FIPS module if they wish to do so.

One possibility that springs to mind is, if there are algorithms that are FIPS validatable, but we as a project are not planning to validate them, then we could add them to the FIPS module anyway and conditionally compile them in or out base on some config parameter. For us that config parameter would be disabled by default, but it gives downstream users the ability to configure which algorithms are "in" or "out" of scope without having to change any source code.

simo5 commented:
> > If the answer is "no" then we should make it as easy as possible for downstream users to pull that code into the FIPS module if they wish to do so.
> 
> One possibility that springs to mind is, if there are algorithms that are FIPS validatable, but we as a project are not planning to validate them, then we could add them to the FIPS module anyway and conditionally compile them in or out base on some config parameter. For us that config parameter would be disabled by default, but it gives downstream users the ability to configure which algorithms are "in" or "out" of scope without having to change any source code.

This sounds entirely reasonable to me.
I do not know if you *should* FIPS validate certain algorithms or not, in general I have the feeling that you should provide a way to make it simple for downstreams to select their list of algorithms.
When you go through valdiation you do not necessarily want to spend the time and money to validate stuff you do not use in your own product (be that a linux distribution or a proprietary gadget that embed opnessl), you generally want to validate only the modules you actively use. So having a way to select a list of modules w/o extensive downpatching would be pretty useful.

Perhaps in the longer term you could provide a fips-algs.txt file that can then generate/compose the fips provider with only the algorithms listed there ?

Aside:
do all fips-able algoritms need to be in common/ ?
If so should I provide a patch to move SSHKDF there? (We do validate SSHKDF in our FIPS certification).

mattcaswell commented:
> do all fips-able algoritms need to be in common/ ?
If so should I provide a patch to move SSHKDF there? (We do validate SSHKDF in our FIPS certification).

IMO we should have all fips-able (nice word) algorithms in "common". But I'd wait a bit for others to express their view before creating PRs.

simo5 commented:
@t8m you may want to chime in on the FIPS angle, or just be informed of this discussion in general :-)

levitte commented:
@simo5, it's entirely possible I got som things wrong and am confused, which is why I ask the questions, to try to bring clarity.
You did talk about *your* lab, which led me to believe that you already had some kind of FIPS "module". However, now that you have mentioned trying to move to have all things crypto being provided by openssl, this makes much more sense to me.

simo5 commented:
@levitte "our lab" is the lab we hire to do FIPS validation of our crypto modules, OpenSSL is one of our crypto modules, which we need to validate, HTH.

levitte commented:
Thanks, that made perfect sense.

frozencemetery commented:
Okay, update ready for further review.

- Rebased since master moved
- Fixed missing EVP_MAC_free()
- Work around missing htobe32()
- Passing CI now.
- Fixed the `strlen()` invocations in the man page.  Please note that the SSKDF man page has this same issue; I have not fixed it.
- Fixed provctx vs. libctx naming.  sskdf also has this issue; I have not fixed it.
- Added a MAC parameter.  Currently still only supports HMAC.
  - Add PROV_R_MISSING_MAC
- Fixed querying the KDF size (I think)
- @levitte, updated mac/digest handling - please recheck since I'm not totally confident it's what you're looking for :)

I've avoided making changes to which provider this lives in until that conversation is completed.

levitte commented on providers/common/kdfs/kbkdf.c:
```diff
@@ -0,0 +1,332 @@
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM mparams[3];
+
+    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, libctx))
```
If you use `ossl_prov_macctx_load_from_params`, you can throw away `ctx->digest` and `ctx->mac`. Basically, that routine will set up the MAC context for you with the MAC it finds and feed it the appropriate digest setup params.

To then check that you got the right MAC, all you have to do is this:

``` C
    if (EVP_MAC_is_a(EVP_MAC_CTX_mac(ctx->ctx_init), OSSL_MAC_NAME_HMAC)) 
        .... 
```

frozencemetery replied:
Sorry, still confused - there's no such function `ossl_prov_macctx_load_from_params()`.  I suppose I could implement it - is that what you're suggesting?

frozencemetery replied:
Ah, I see you added it in #9946.  I'll rebase on top of that.

slontis commented:
> IMO we should have all fips-able (nice word) algorithms in "common". But I'd wait a bit for others to express their view before creating PRs.

I agree with this. The ciphers have all been written this way. Try to write your code so that it could be easily moved to another folder (default or legacy). Something that is a fips algorithm now might not be so at some point in the future.

frozencemetery commented:
Rebased on top of #9946 to take advantage of `ssl_prov_macctx_load_from_params()`; please ignore extra commits.

paulidale commented:
It might be worthwhile have a discussion with your lab about dependent modules.
My understanding (which could be completely wrong) is that you can define your FIPS boundary so that it encompasses your module and another and that your module can use the other without breaking FIPS-ness.  In this case, OpenSSL would keep its FIPS module as is and a Kerberos module added.  The latter would be validated dependent on the former.  Anyway, I'm not a FIPS lawyer so there could be show stoppers I don't know about.  But it might be worth a question.


The OpenSSL validation is intended to be a **source** validation just like the old FOM.  This should allow anyone to compile in the appropriate manner and claim compliance.  Binary validations are easier but not the only ones possible.


Finally, we could ask our lab if there would be any additional costs involved in included these algorithms -- I believe that they are all vendor affirmed (i.e. no testing required) and if so, the extra work for the lab is listing them as included and usable in the module but not formally validated on the certificate and in the security policy.


I'd almost be tempted to make a separate provider to hold these until a decision can be made either way.  More so we don't lose track of which are in this category than anything else.  Thus far these two KDF plus SSHKDF I think.

t8m commented:
> @t8m you may want to chime in on the FIPS angle, or just be informed of this discussion in general :-)

@simo5 I follow this discussion and I think all that you've said summarizes our position quite nicely.

@paulidale We will probably not use the openssl source FIPS validation so we will have an option to adjust the fips provider to suit our needs. However it would be helpful if the source code organization of the upstream OpenSSL code already easily allowed to add further algorithms from the OpenSSL codebase to the fips provider. I suppose it should be easily possible to make it done in a way that the original intended scope of the upstream OpenSSL fips provider is not expanded.

paulidale commented:
@t8m, not expanding the original FIPS provider seems like a good way to go.  E.g. might it be easier to have a `providers/redhat-fips` directory and pull what you need into there?

t8m commented:
Yep, that should work, although I would not use redhat in the directory name - perhaps something like fips-approved or fips-additional would be better.

richsalz commented:
Perhaps OpenSSL should have a vendor or third-party directory in providers and tooling to automatically pick up sub-directories under that?

levitte commented:
> Perhaps OpenSSL should have a vendor or third-party directory in providers and tooling to automatically pick up sub-directories under that?

Generally speaking, I think that's the wrong move.  Third parties should be able to build their own provider modules and distribute them on their own.

levitte commented:
(heh, on a semi-serious note, I kinda like "FIPSish")

simo5 commented:
So a quick note, I do not like much the directory approach, as it require to unnecessarily move code around (if I understand how it works).
I would rather have a distinction between algorithm implementation (and where it lives) and how a provide is composed.
FIPS is a sort of "virtual" provider in upstream OpenSSL as you are not creating distinct code for it (mostly), but simply picking and choosing from a menu of algorithms.
Given that's what you are doing, and given this choice may/will change over time I feel like it would be easier to work with the project if you could crate providers via a selection mechanism that can pick algorithms from where they are without necessarily moving them around.
Of course there will also *always* be the option to create a directory, copy a bunch of code and say that is a provider, it just seem like having that as the only option is a bit constraining.

simo5 commented:
> It might be worthwhile have a discussion with your lab about dependent modules.

So first a language question, when you say modules, are you using the FIPS "cryptography module" definition, or do you refer to openssl providers ?

> My understanding (which could be completely wrong) is that you can define your FIPS boundary so that it encompasses your module and another and that your module can use the other without breaking FIPS-ness. In this case, OpenSSL would keep its FIPS module as is and a Kerberos module added.

Here it seems you are talking about providers so I will assume that's the meaning in the rest of the comment.

> The latter would be validated dependent on the former. Anyway, I'm not a FIPS lawyer so there could be show stoppers I don't know about. But it might be worth a question.

So while you can have dependent modules, at least till now, we've been conservative and assumed that you can only use specific APIs that have been "validated as entrypoints" in the cryptographic module. If we are talking about two separate OpenSSL modules, this distinction will be hard to make, because the kerberos provider will use a lot of internal interfaces of OpenSSL. That said in our case it is probably a moot point because we evaluate the whole of OpenSSL as a crytpo module so the "upstream FIPS module" and an hypotetical "dowstream FIPS module" will be validated at the same time and no dependency exists, they will all be part of the same crytpo module for fIPS purposes and specific algorithm will be claimed and CVMP certificates obatined.
 
> The OpenSSL validation is intended to be a **source** validation just like the old FOM. This should allow anyone to compile in the appropriate manner and claim compliance. Binary validations are easier but not the only ones possible.

Source validation is useful as a starting point, but it is not sufficient. Vendor affirmation go only so far, and with FOM you can only do Vendor affirmation. However our customers really need CAVS (or ACVP) and CMVP certificates. We can vendor affirm that dependent userspace then uses openssl as a crypto module and does not require certification. But the core crypto really needs actual certificates in our case.

> Finally, we could ask our lab if there would be any additional costs involved in included these algorithms -- I believe that they are all vendor affirmed (i.e. no testing required) and if so, the extra work for the lab is listing them as included and usable in the module but not formally validated on the certificate and in the security policy.

I am a bit fuzzy on what you want to claim an why, from my POV what really matters to a vendor that is going to go through actual testing and full FIPS validation is that we can pick and chose which algorithms we want to approve in our FIPS module without that causing huge code churn (like moving around entire files). Code churn makes it a lot harder and error prone to later backport changes from upstream on maintenance releases. So minimizing code movement where reasonable is definitely a big plus for us.

> I'd almost be tempted to make a separate provider to hold these until a decision can be made either way. More so we don't lose track of which are in this category than anything else. Thus far these two KDF plus SSHKDF I think.

When you say we make a separate provider what does it mean exactly?
We do not have to put them in any special provider if they do not go in the FIPS provider, they can stay in the default provider?

Just to be clear, for us where the code lives makes no difference from the pov of FIPS validation. What makes a difference is whether the algorithm will be available for use or not when the system is switched to FIPS mode. In that mode only the approved algorithms should be available to userspace applications, unless the application claim that the use of an algorithm is for "non security" purposes.
So in the end I am not entirely sure a distinct FIPS provider adds a ton of value to us (it may be different in more targeted environments of course), we still want to allow applications that know what they are doing to call into unapproved algorithms.

As an example we have applications that use MD5 just for name normalization, there use is not for security, just to implement existing standards where they decided to use a Hash function to normalize names down to a known length, and where potential collisions are a non issue.
If we create a FIPS provider where Md5 is obviously exclude would it mean applications suddenly won't be able to access this hash at all ? Sorry if this is an answered question already.

Hope I clarified our PoV though.

richsalz commented:
My guess is that there are downstream users of openssl who will work the main repo and add their own files.  Akamai does this.  It is *possible* we will add to the core provider, but probably unlikely. I do think that this is a worthwhile use-case to support.

Probably open a new issue for this discussion tho?

simo5 commented:
@richsalz I am happy to open an issue, but not entirely sure how to frame it I am afraid, I need to confess the providers stuff and their ramification is still not clear enough in my mind to write a sensible issue topic and description :-/

levitte commented:
> So first a language question, when you say modules, are you using the FIPS "cryptography module" definition, or do you refer to openssl providers ?

In reference to our FIPS provider module (i.e. everything that gets built into providers/fips.so), that *is* also our FIPS "cryptography module" for OpenSSL 3.0.  So in our terms, they are one and the same.

Speaking of language, you wrote that you view all of OpenSSL as krb5's crypto module, do you mean that in FIPS "cryptography module" terms, i.e. are you including libcrypto in that?

If I hear it all correctly, it seems that we have differing views on where the FIPS "cryptography module" boundary goes.

Caveat: I may be wrong or confused, or both.  I expect to be duly corrected.

simo5 commented:
> > So first a language question, when you say modules, are you using the FIPS "cryptography module" definition, or do you refer to openssl providers ?
> 
> In reference to our FIPS provider module (i.e. everything that gets built into providers/fips.so), that _is_ also our FIPS "cryptography module" for OpenSSL 3.0. So in our terms, they are one and the same.
> 
> Speaking of language, you wrote that you view all of OpenSSL as krb5's crypto module, do you mean that in FIPS "cryptography module" terms, i.e. are you including libcrypto in that?
> 
> If I hear it all correctly, it seems that we have differing views on where the FIPS "cryptography module" boundary goes.
> 
> Caveat: I may be wrong or confused, or both. I expect to be duly corrected.

It is indeed confusing especially because we do make a distinction about what is the boundary in some cases, but indeed we do consider libcrypto to be the crypto module, although not all of the exported functions are approved for use in FIPS mode.

In the end it is a matter of what you send through the lab and more down to earth what is the binary you produce. Up to OpenSSL 1.1.1 libcrypto.so is the binary we test, so that's the module.
If, in OpenSSL 3.0, providers will be built out to their own separate binary (.so objects) then we could restrict the definition to just that part I guess, we just haven't thought as far as OpenSSL v3 wrt these details yet.

paulidale commented:
I was writing about FIPS modules not providers.  Still, this quickly gets into a FIPS morass pretty quickly.  If you are validating all of libcrypto, there shouldn't be a problem.

I'll note a couple of things:

1. People not happy with a vendor affirmation, can pay a lab to clone the FIPS validation.  This isn't suitable for you but many will do so.
2. Accessing non-FIPS functionality will be possible even when running in FIPS mode.  This has been part of the design since the beginning.  It will require an explicit choice by the caller to access.
3. I like the idea of putting all the algorithm sources in one place and have providers pick which pieces they want.  I'm not sure how this interacts with the current build framework.  Perhaps something to discuss on the project list.

simo5 commented:
@levitte is there anything else that needs to be done here (besides rebasing on master)

frozencemetery commented:
Rebased onto master.  (Trivial rebase; not sure what github's problem was.)

levitte commented:
At this point, I like what I see.  However, it's late evening and I'm quite tired, so I'll have a second read tomorrow morning.  Also, is add this to the FIPS module still a controversy?

mattcaswell commented:
> Also, is add this to the FIPS module still a controversy?

I think as far as this PR is concerned we just add it. There does need to be a separate discussion about whether these algorithms are pulled into our own validation or not, and whether we provide downstream users with a mechanism for "picking and choosing" which algorithms exist within the module. We don't need to decide those things right now. In the worst case scenario, at some point before the release of 3.0, we will have to remove these algorithms from our fips module but keep them in default. I don't think it will come to that though.

levitte commented:
Merged.

a39bc4404baa4e065d01efe829a1f26eba737049 [KDF] Add KBKDF implementation for counter-mode HMAC

levitte commented:
We already have a mechanism for "picking and choosing", in form of `no-` configuration options
